### DISCUSSION TITLE
How to download raster data from WCS

### DISCUSSION THREAD

**Discussion by dominicroye** [eblondel/ows4R#104] [OGC WCS] [answered]
Hi!

This is a great package initiative, but I honestly don't get how to access and download raster data via a WCS server. 

I am trying so with the following https://wcs-mds.idee.es/mds?version=2.0.1&request=GetCapabilities&service=WCS in order to download mdsn_e025 data with a bbox for a small area. 

Could you give me any idea of how to proceed? It's not very clear to me. 

I understand to start in the following way: 

`CSW <- CSWClient$new("https://wcs-mds.idee.es/mds", "2.0.1", logger = "INFO")`

Thank you!

Best,

**Comment by eblondel** [comment, marked as answer]
I've tried the following code (see additional notes in the comments...).

At first, you should use the `WCSClient` to interrogate Web Coverage Service, not the `CSWClient` which is for geographic metadata cataloguing. 

Regarding the data access, this server seems to be limited to very small data coverages.

```R
require(ows4R)

#get WCS client connection (be careful, it's WCS, not CSW)
#WCS = Web Coverage Service (for raster data access)
#CSW = Catalogue Service for the Web (for geographic metadata cataloguing)
WCS = WCSClient$new(
  url = "https://wcs-mds.idee.es/mds",
  serviceVersion = "2.0.1",
  logger = "INFO"
)

#find the coverage
cov = WCS$capabilities$findCoverageSummaryById("mdsn_e025")

#get data dimensions (here's a spatial 2D coverage)
cov$getDimensions()

#get data
covdata = cov$getCoverage()
#this doesn't work
#If you look at output of request on the web, you will see this error:
#"Raster size out of range, width and height of resulting coverage must be no more than MAXSIZE=4096"
#(This seems to authorize very small data coverages...)

#trying with one bounding box
covdata = cov$getCoverage( bbox = OWSUtils$toBBOX(xmin = -100,xmax = 1000,ymin = 3500000,ymax = 3500200))

```
The request works, but so far I see the cov data is empty (all values equal to zero). This might be due to the extent i've chosen, not sure. If you know some bounding boxes you want to apply, can you try it?

**Reply by eblondel** [reply]
btw @dominicroye I will look further into this service provider, and test it further with ows4R. The WCS client is quite new in ows4R, and I don't exclude it needs consolidation, for this more WCS services are tested, the better is.

**Comment by dominicroye** [comment]
Thank you very much, Emmanuel! It's working fine. The data are 2.5m x 2.5m of building heights in Spain, this is why you need a very small area. 

`
library(sf)
library(mapSpain)

bbx <- esp_get_munic(munic = "Coruña, A") |> st_transform(25830) |> st_bbox()

#trying with one bounding box
covdata = cov$getCoverage(bbox = OWSUtils$toBBOX(xmin = bbx[1], xmax = bbx[3], ymin = bbx[2], ymax = bbx[4]))

plot(covdata)
`
![Rplot](https://user-images.githubusercontent.com/42300133/222893630-db782363-a80c-4ba9-ab1c-0c56714d8def.png)

**Reply by eblondel** [reply]
Great! Nice to see it works!

**Comment by jorgesainzdeaja** [comment]
Hi!, 

have a similar problem here, I have tried the code above and works perfectly, but when trying to get VOM data from an UK WCS service I am getting the following error:

```
Error in if (axisLatIdx < axisLonIdx) bbox <- rbind(bbox[2, ], bbox[1,  : 
  argument is of length zero
```

```
focus_bbox_27700
    xmin     ymin     xmax     ymax 
519713.3 175841.0 522783.3 178911.0 

VOM <- WCSClient$new("https://environment.data.gov.uk/spatialdata/vegetation-object-model/wcs", "2.0.1", logger = "INFO")

VOM_cov <- VOM$capabilities$findCoverageSummaryById("ecae3bef-1e1d-4051-887b-9dc613c928ec__Vegetation_Object_Model_Elevation_2022", exact = F)

VOM_data <- VOM_cov$getCoverage(bbox = OWSUtils$toBBOX(focus_bbox_27700[1], focus_bbox_27700[3], focus_bbox_27700[2], focus_bbox_27700[4]))
```

What I am doing wrong??, tested to load the WCS layer on QGIS and it loads just right
